### ISSUE TITLE
Uncaught ReferenceError: process is not defined | Using env variables with Vite

### ISSUE THREAD

**Issue by AdityaSaroj** [remix-run/remix#8448]
### Reproduction

I checked the console and the only error I have is this:
```
Uncaught ReferenceError: process is not defined
    at node_modules/util/util.js (@remix-run_node.js?v=2071f225:764:5)
    at __require (chunk-WXXH56N5.js?v=2071f225:12:50)
    at node_modules/stream-slice/index.js (@remix-run_node.js?v=2071f225:5672:16)
    at __require (chunk-WXXH56N5.js?v=2071f225:12:50)
    at node_modules/@remix-run/node/dist/upload/fileUploadHandler.js (@remix-run_node.js?v=2071f225:5869:23)
    at __require (chunk-WXXH56N5.js?v=2071f225:12:50)
    at node_modules/@remix-run/node/dist/index.js (@remix-run_node.js?v=2071f225:6057:29)
    at __require (chunk-WXXH56N5.js?v=2071f225:12:50)
    at @remix-run_node.js?v=2071f225:6159:16
```
vite.config.ts:
```
import { unstable_vitePlugin as remix } from "@remix-run/dev";
import { ConfigEnv, defineConfig, loadEnv } from "vite";
import tsconfigPaths from "vite-tsconfig-paths";

export default ({ mode }: ConfigEnv) => {
  // Here we add env vars from .env files to process.env.
  // Note the last arg is a blank string so that all env vars
  // are loaded, not just those starting with "VITE_"
  Object.assign(process.env, loadEnv(mode, process.cwd(), ""));

  return defineConfig({
    plugins: [
      remix({
        ignoredRouteFiles: ["**/.*"],
        serverModuleFormat: "cjs",
      }),
      tsconfigPaths(),
    ],
  });
};
```
How I'm accessing the env variables is as per the documentation: https://remix.run/docs/en/main/guides/envvars

root.tsx:
```
export async function loader({ request }: LoaderFunctionArgs) {
  const isAnonymous = await isAnonymousSession(request);

  if (isAnonymous) {
    return response.ok(
      {
        env: getBrowserEnv(),
        email: null,
        userTier: null,
      },
      { authSession: null }
    );
  }

  const authSession = await requireAuthSession(request);
  const { userId, email } = authSession;

  try {
    const userTier = await getUserTier(userId);

    return response.ok(
      {
        env: getBrowserEnv(),
        email,
        userTier,
      },
      { authSession }
    );
  } catch (cause) {
    throw response.error(cause, { authSession });
  }
}
export default function App() {
  const { env } = useLoaderData<typeof loader>();

  return (
    <html className="h-full" lang="en">
      <head>
        <Meta />
        <meta charSet="utf-8" />
        <meta
          name="viewport"
          content="width=device-width,initial-scale=1.0,maximum-scale=1.0"
        />
        <Links />
      </head>
      <body className="h-full">
        <Content />
        <ScrollRestoration />
        <script
          dangerouslySetInnerHTML={{
            __html: `window.env = ${JSON.stringify(env)}`,
          }}
        />
        <Scripts />
        <LiveReload />
      </body>
    </html>
  );
}
```
env.ts:
```
import { Currency } from "@prisma/client";
import { z } from "zod";

import { OneMixError } from "~/utils/error";
import { isBrowser } from "./misc";

declare global {
  interface Window {
    env: {
      SUPABASE_URL: string;
      SUPABASE_ANON_PUBLIC: string;
      DEFAULT_CURRENCY: Currency;
    };
  }
}

declare global {
  namespace NodeJS {
    interface ProcessEnv {
      SUPABASE_URL: string;
      SUPABASE_SERVICE_ROLE: string;
      SERVER_URL: string;
      SUPABASE_ANON_PUBLIC: string;
      SESSION_SECRET: string;
      STRIPE_SECRET_KEY: string;
      STRIPE_ENDPOINT_SECRET: string;
      DEFAULT_CURRENCY: Currency;
    }
  }
}

type EnvOptions = {
  isSecret?: boolean;
  isRequired?: boolean;
};
function getEnv(
  name: string,
  { isRequired, isSecret }: EnvOptions = { isSecret: true, isRequired: true }
) {
  if (isBrowser && isSecret) return "";

  const source = (isBrowser ? window.env : process.env) ?? {};

  const value = source[name as keyof typeof source] || "";

  if (!value && isRequired) {
    throw new OneMixError({
      message: `Env "${name}" is not set`,
    });
  }

  return value;
}

/**
 * Server env
 */
export const SERVER_URL = getEnv("SERVER_URL");
export const SUPABASE_SERVICE_ROLE = getEnv("SUPABASE_SERVICE_ROLE");
export const SESSION_SECRET = getEnv("SESSION_SECRET");
export const STRIPE_SECRET_KEY = getEnv("STRIPE_SECRET_KEY");
export const STRIPE_ENDPOINT_SECRET = getEnv("STRIPE_ENDPOINT_SECRET");

/**
 * Shared envs
 */
export const NODE_ENV = getEnv("NODE_ENV", {
  isSecret: false,
  isRequired: false,
});
export const SUPABASE_URL = getEnv("SUPABASE_URL", { isSecret: false });
export const SUPABASE_ANON_PUBLIC = getEnv("SUPABASE_ANON_PUBLIC", {
  isSecret: false,
});
export const DEFAULT_CURRENCY = z
  .nativeEnum(Currency)
  .parse(getEnv("DEFAULT_CURRENCY", { isSecret: false }));

export function getBrowserEnv() {
  return {
    SUPABASE_URL,
    SUPABASE_ANON_PUBLIC,
    DEFAULT_CURRENCY,
  };
}
```

### System Info

```shell
System:
    OS: Windows 11 10.0.22631
    CPU: (12) x64 Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
    Memory: 1.00 GB / 7.86 GB
  Binaries:
    Node: 20.9.0 - D:\Node\node.EXE
    npm: 10.2.4 - ~\AppData\Roaming\npm\npm.CMD
    pnpm: 8.11.0 - ~\AppData\Local\pnpm\pnpm.EXE
  Browsers:
    Edge: Chromium (120.0.2210.121)
    Internet Explorer: 11.0.22621.1
  npmPackages:
    @remix-run/dev: ^2.4.0 => 2.4.1 
    @remix-run/node: ^2.4.0 => 2.4.1 
    @remix-run/react: ^2.4.0 => 2.4.1 
    @remix-run/serve: ^2.4.0 => 2.4.1 
    vite: ^5.0.0 => 5.0.10
```

### Used Package Manager

npm

### Expected Behavior

There should be a process is not defined error.

### Actual Behavior

There is a process is not defined error in the browser console. As a side effect of this, the onClick function for any of my buttons is not working.

**Comment by Pape98** [comment]
Literally having the same exact issue as well. Issue is happening for me with this line: `import { json, redirect } from "@remix-run/node`

**Comment by AdityaSaroj** [comment]
Any workaround, @Pape98?

**Comment by pcattori** [comment]
You may be mixing server-only code and client-safe code in the same module (looks like `env.ts` is doing this on purpose).

See our new docs on how to fix/workaround this issue: https://remix.run/docs/en/main/future/vite#server-code-errors-in-browser-during-development

**Comment by toddheslin** [comment]
For anyone coming across this problem, I had a similar issue where I had a common `utils.ts` file with some server + client utils. I separated them to a `utils.ts` and `utils.server.ts` such that the latter was the only one that was importing:
```ts
import { json } from '@remix-run/node'
```
This seemed to have fixed the issue. Alternatively, I could have used the `vite-env-only` package to have one file and ensure the individual functions are nominated as client or server only. Two files made more sense for my project.

**Comment by pcattori** [comment]
Closing as we have dedicated docs for [splitting up client and server code](https://remix.run/docs/en/main/future/vite#splitting-up-client-and-server-code) as well as a section within [Vite > Troubleshooting](https://remix.run/docs/en/main/future/vite#server-code-errors-in-browser-during-development) for this exact issue.

**Comment by 684efs3** [comment]
Is there a way for the error message to track down where the issue is?

**Comment by pcattori** [comment]
> Is there a way for the error message to track down where the issue is?

The error message comes from JS itself, not Remix. If Remix tries to handle the error, it would prevent users from setting up client-safe polyfills for the same imports so not much we can do without a substantially more complex approach.

**Comment by cha0s** [comment]
Say you have a file `cookie.js`:

```js
import { createCookie } from "@remix-run/node";
export const prefs = createCookie("prefs");
```

Trying to use it like:

```js
import {prefs} from './cookie';

export async function loader({request}) {
  const cookieHeader = request.headers.get("Cookie");
  const cookie = (await prefs.parse(cookieHeader)) || {};
  return json({
    // ...
  });
}

```

will throw. However, **the following dynamic import will work**:

```js
export async function loader({request}) {
  const {prefs} = await import('./cookie');
  const cookieHeader = request.headers.get("Cookie");
  const cookie = (await prefs.parse(cookieHeader)) || {};
  return json({
    // ...
  });
}
```

The compiler will shake out the dynamic import for the client bundle and everything will be peachy. Hope this helps someone!

**Comment by spacepolice10** [comment]
> Closing as we have dedicated docs for [splitting up client and server code](https://remix.run/docs/en/main/future/vite#splitting-up-client-and-server-code) as well as a section within [Vite > Troubleshooting](https://remix.run/docs/en/main/future/vite#server-code-errors-in-browser-during-development) for this exact issue.

So it is a feature and now I can't set up a universal fetch? That feels like an extreme downgrade and it really counterintuitive considering recent changes in client data loading. Any chances that everything will work as normal or should I think about ignoring Vite and stay with remix-serve?

**Comment by spacepolice10** [comment]
@pcattori let's say I have a function `customFetch`. It is a fetch wrapper that is retrieving cookie and makes authorized calls to the server. This `customFetch` determines the environment using `typeof window == 'undefined'` check & depending on the environment grabs cookies using either **remix/node function `createCookie`** or **client-side library with the same functionality**.  If I understand correct now I have to explicitly create `customFetch.server` & `customFetch.client`. But if I import `customFetch.server` to any function that is not marked as `.server` I will get an error and the code execution will stop. So I would assume that I must now create the double set of all the queries that I make for the server-side and client-side explicitly or pass all the data necessary to make a query (e.g. taking tokens out of cookies) manually

**Comment by pcattori** [comment]
@spacepolice10 check out [vite-env-only](https://github.com/pcattori/vite-env-only). Then you can do something like:

```ts
// customFetch.ts
import { clientOnly$, serverOnly$ } from "vite-env-only"

const clientFetch = clientOnly$(() => {
  // client-only code
})

const serverFetch = serverOnly$(() => {
  // server-only code
})

export const customFetch = () => {
  if (typeof window === 'undefined') return serverFetch()
  return clientFetch()
}
```

**Comment by spacepolice10** [comment]
@pcattori Thank you for spending your time on my problem and providing an example. I'll check out this approach as soon as possible

**Comment by fabregas4** [comment]
@Pape98 @toddheslin 

I'm a newbie so I might be off-track, but I have struggled with what seems a very similar issue over the last 24 hours. I finally resolved it (I think), at least for me.

I was trying to `import { json } from '@remix-run/node'` as well.

I was getting that same `process` var error in browser console.

I tried moving the import to a utils.server.js and then importing that from my route. This seemed to clear the browser console error, but now became a server console error "Internal server error: Server-only module referenced by client".

What I eventually discovered is you have to USE the `json `method within your loader or action. If you do this, the error should clear, regardless of whether you import the package directly or via a .server file.

I was tripped up because I was actually importing both `json` and `redirect`, but only using one or the other in the action. The error only cleared when both were being used, or the one that wasn't being used, was removed from the import.

Hope that makes sense and hope it helps.

Basically my conclusion is that if you import methods without using, vite/remix assumes they are referenced by the client, whereas if you use them within loader/action, then it understands they are only referenced by the server. Perhaps someone from the team can comment?

**Comment by toddheslin** [comment]
@fabregas4 were you using `json` or `redirect` inside some other module (not inside the action or loader) that wasn't a `xxxx.server.ts` file? I would think that it would be fine if it was a `xxxx.server.ts`, although I personally only use those remix functions directly inside loaders/actions and move business logic/validation to other modules. 

This does sound similar to where I had a module that was neither `.client` or `.server` and I was relying on the bundler to only grab the dependencies that were relevant for the function I was importing. I think the way we do it now is much better and more explicit but I can see some migration pains for existing apps that relied on a bit of bundler magic to figure this out.

**Comment by fabregas4** [comment]
@toddheslin I don't believe I was using `json` or `redirect` anywhere else.

The errors cleared just by using these methods within the loader of the route which had the import, without touching any other code within the app.

**Comment by toddheslin** [comment]
OK, but if you say 'just by using these in the loader', this means they were in some other file/module elsewhere before right? I'm curious where it was causing problems before.

**Comment by LukeAskew** [comment]
@toddheslin I had the same issue as @fabregas4.

The error occurs if you import either `json` or `redirect` and don't actually use them in your loader/action. I hit this in development when I implemented a redirect, but then backed it out without actually removing the import as well.

**Comment by spacepolice10** [comment]
> @spacepolice10 check out [vite-env-only](https://github.com/pcattori/vite-env-only). Then you can do something like:
> 
> ```ts
> // customFetch.ts
> import { clientOnly$, serverOnly$ } from "vite-env-only"
> 
> const clientFetch = clientOnly$(() => {
>   // client-only code
> })
> 
> const serverFetch = serverOnly$(() => {
>   // server-only code
> })
> 
> export const customFetch = () => {
>   if (typeof window === 'undefined') return serverFetch()
>   return clientFetch()
> }
> ```

@pcattori I tried this one. Unfortunately the error persists client-side. My app is more useful now but I don't see RemixDevTools or QueryDevTools, I can't manipulate popovers and other clientSide dynamic elements. If I build the app it considers all the queries made by Tanstack to be routes and awlays returns something like `INVALID_URL: "{endpoint that tanstack tried to reach}"` and breaks the application (server provides no errors). And I can't event chase the error because it says that this one appears in a file called `fileUploadeHandler` in `node_modules/remix`. It is cryptic

**Comment by markhker** [comment]
In my case I was importing a cookie into a file containing several functions, some of those functions were being used server side and some of them client side, the cookie was only used in the server side functions. Wrapping the server side functions (the ones using the cookie) inside a `serverOnly$` function solved the problem, an alternative would be separating the functions into multiple files. Hope this helps someone.

**Comment by kevinvugts** [comment]
> @toddheslin I had the same issue as @fabregas4.
> 
> The error occurs if you import either `json` or `redirect` and don't actually use them in your loader/action. I hit this in development when I implemented a redirect, but then backed it out without actually removing the import as well.

thank you so much for this!

**Comment by gao-sun** [comment]
my case may be rare, but hopefully helpful to somebody. i was putting `loader` and `meta` in the same file, and use `redirect` in the loader. the code looks like:

```ts
// loader.ts
import { redirect } from '@remix-run/node';

export const loader = async () => {
  // ...
  return redirect('/somewhere');
};

export const meta = () => {
  // ...
};
```

and

```ts
// root.ts

export { loader, meta } from './loader';

// ...
```

the browser will show the "process is not defined" error. however, when i remove the export in the `root.ts`:

```ts
// root.ts

export { loader } from './loader';
```

it works.

i ended up with returning `new Response()` to redirect instead of using the util:

```ts
// loader.ts

export const loader = async () => {
  // ...
  return new Response(null, {
    status: 301,
      headers: {
        Location: '/somewhere',
      },
  });
};

// ...
```
